### PR TITLE
Convert benchmarks to new format, expand normalization benchmarks to fp16.

### PR DESCRIPTION
**Pull request by csarofeen**

### CODE REVIEW

File: benchmarks/cpp/nvfuser/batch_norm.cpp [hunk 1]
```diff
@@ -60,41 +58,84 @@ static void BatchNorm(benchmark::State& benchmark_state) {
...
+    benchmark_state.SetIterationTime(
+        executor_instance->kernelTimeMs() / 1000.0);
   }
+  // Sync everything up before we're finished, don't want to run ahead on the
+  // cpu while benchmarking.
+  cudaDeviceSynchronize();
+
+  benchmark_state.SetBytesProcessed(
+      (int64_t(benchmark_state.iterations()) *
+       (2 * at_x.numel() + 2 * at_weight.numel()) *
```
Comment (rdspring1) [inline benchmarks/cpp/nvfuser/batch_norm.cpp]: nit: for better clarity:
`2 * at_weight.numel()` -> `at_weight.numel() + at_bias.numel()`
`2 * at_weight.numel()` -> `at_run_mean.numel() + at_run_var.numel()`

File: benchmarks/cpp/nvfuser/batch_norm.cpp [hunk 2]
```diff
@@ -134,16 +178,60 @@ static void BatchNorm_Baseline(benchmark::State& benchmark_state) {
     benchmark_state.SetIterationTime(timer.elapsed() / 1000.0);
     cudaDeviceSynchronize();
   }
+
+  benchmark_state.SetBytesProcessed(
+      (int64_t(benchmark_state.iterations()) *
+       (2 * at_x.numel() + 2 * at_weight.numel()) *
```
Comment (rdspring1) [inline benchmarks/cpp/nvfuser/batch_norm.cpp]: nit: for better clarity
`2 * at_weight.numel()` -> `at_weight.numel() + at_bias.numel()`
`2 * at_weight.numel()` -> `at_run_mean.numel() + at_run_var.numel()`

File: benchmarks/cpp/nvfuser/instance_norm.cpp [hunk 3]
```diff
@@ -141,23 +89,34 @@ static void InstanceNorm_NvFuser(
...
+  // Sync everything up before we start
   cudaDeviceSynchronize();
   for (auto _ : benchmark_state) {
-    CudaKernelTimer timer;
-    outputs = fec.runFusionWithInputs(inputs);
-    benchmark_state.SetIterationTime(timer.elapsed() / 1000.0);
-    cudaDeviceSynchronize();
+    auto cg_outputs = fusion_executor_cache->runFusionWithInputs(aten_inputs);
+    benchmark_state.SetIterationTime(
+        executor_instance->kernelTimeMs() / 1000.0);
```
Comment (rdspring1) [inline benchmarks/cpp/nvfuser/instance_norm.cpp]: InstanceNorm requires two kernels. I think `fusion_executor_cache->getMostRecentExecutorInfo()` will return the executor of the last run kernel, so we would only get the running time for the last kernel.

File: benchmarks/cpp/nvfuser/layer_norm.cpp [hunk 4]
```diff
@@ -16,58 +16,97 @@ using namespace torch::jit::fuser::cuda;
...
+    benchmark_state.SetIterationTime(
+        executor_instance->kernelTimeMs() / 1000.0);
   }
+  // Sync everything up before we're finished, don't want to run ahead on the
+  // cpu while benchmarking.
+  cudaDeviceSynchronize();
+
+  benchmark_state.SetBytesProcessed(
+      int64_t(benchmark_state.iterations()) * 2 *
+      (input.numel() + input_shape[0]) * int64_t(dataTypeSize(dtype)));
```
Comment (rdspring1) [inline benchmarks/cpp/nvfuser/layer_norm.cpp]: I made a mistake with `input_shape[0]`. 

I'd replace `2 *(input.numel() + input_shape[0])` with `2 * input.numel() + weight.numel() + bias.numel()`

File: benchmarks/cpp/nvfuser/softmax.cpp [hunk 5]
```diff
@@ -182,21 +276,216 @@ static void Softmax_Dropout_Baseline(benchmark::State& benchmark_state) {
     benchmark_state.SetIterationTime(timer.elapsed() / 1000.0);
     cudaDeviceSynchronize();
   }
+
+  benchmark_state.SetBytesProcessed(
+      int64_t(benchmark_state.iterations()) * 6 * 256 * 12 * 100 *
```
Comment (rdspring1) [inline benchmarks/cpp/nvfuser/softmax.cpp]: Maybe replace the hardcoded size `256 * 12 * 100 * benchmark_state.range(0)` with a constant or `attention_scores.numel()`

Comment (rdspring1) [reply benchmarks/cpp/nvfuser/softmax.cpp]: Should this be `5 * input_numel * dataTypeSize(dtype) + input_numel * dataTypeSize(Bool)` where `input_numel = 256 * 12 * 100 * benchmark_state.range(0)`?

File: benchmarks/cpp/nvfuser/softmax.cpp [hunk 6]
```diff
@@ -97,62 +141,111 @@ static void Softmax_Dropout(benchmark::State& benchmark_state) {
...
+    auto cg_outputs = fusion_executor_cache->runFusionWithInputs(aten_inputs);
+    benchmark_state.SetIterationTime(
+        executor_instance->kernelTimeMs() / 1000.0);
   }
+  // Sync everything up before we're finished, don't want to run ahead on the
+  // cpu while benchmarking.
+  cudaDeviceSynchronize();
+
+  benchmark_state.SetBytesProcessed(
+      int64_t(benchmark_state.iterations()) * 6 * 256 * 12 * 100 *
```
Comment (rdspring1) [inline benchmarks/cpp/nvfuser/softmax.cpp]: Should `SetBytesProcessed` be `5 * kInputNumel * dataTypeSize(dtype) + kInputNumel * dataTypeSize(Bool)` where `kInputNumel = 256 * 12 * 100 * benchmark_state.range(0)`?

**Review by rdspring1 — APPROVED** [review]
LGTM. 

If `benchmark_state.SetIterationTime` is not set, the benchmark can run for a long time, so I used `CudaKernelTimer timer` for segmented fusions.

File: benchmarks/cpp/nvfuser/utils.cpp [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+      clearL2Cache();
+    }
+    // Sync everything up before we're finished, don't want to run ahead on the
+    // cpu while benchmarking.
+    cudaDeviceSynchronize();
+  } else {
+    // Segmented
+    // Sync everything up before we start
+    cudaDeviceSynchronize();
+    CudaKernelTimer timer;
```
Comment (rdspring1) [inline benchmarks/cpp/nvfuser/utils.cpp:93]: I think the compile-time for the fusion is included in the total running time unless `runFusionWithInputs` is run once before timing.
